std/assembly/wasi/index.ts

```diff
@@ -0,0 +1,121 @@
+  }
+  store<u32>(ptr, 0x206E6920); ptr += 4; // ' in '
+  if (fileName !== null) {
+    ptr += String.UTF8.encodeUnsafe(changetype<usize>(fileName), fileName.length, ptr);
+  }
+  store<u8>(ptr++, 0x28); // (
+  var len = decimalCount32(lineNumber); ptr += len;
+  do {
+    let t = lineNumber / 10;
+    store<u8>(--ptr, 0x30 + lineNumber % 10);
+    lineNumber = t;
+  } while (lineNumber); ptr += len;
+  store<u8>(ptr++, 0x3A); // :
+  len = decimalCount32(columnNumber); ptr += len;
+  do {
+    let t = columnNumber / 10;
+    store<u8>(--ptr, 0x30 + columnNumber % 10);
+    columnNumber = t;
+  } while (columnNumber); ptr += len;
+  store<u16>(ptr, 0x0A29); ptr += 2; // )\n
+  changetype<iovec>(iovPtr).buf_len = ptr - bufPtr;
+  fd_write(2, iovPtr, 1, offsetof<iovec>());
+  proc_exit(255);
+}
+
+function trace(
+  message: string,
+  n: i32 = 0,
+  a0: f64 = 0,
+  a1: f64 = 0,
+  a2: f64 = 0,
+  a3: f64 = 0,
+  a4: f64 = 0
+): void {
+  // 0: iov.buf
+  // 4: iov.buf_len
+  // 8: len
+  // 12: buf...
+  var iovPtr = __alloc(offsetof<iovec>() + sizeof<usize>() + 1 + <usize>(max(String.UTF8.byteLength(message), MAX_DOUBLE_LENGTH << 1)), 0);
+  var lenPtr = iovPtr + offsetof<iovec>();
+  var bufPtr = lenPtr + sizeof<usize>();
+  changetype<iovec>(iovPtr).buf = bufPtr;
+  store<u64>(bufPtr, 0x203A6563617274); // 'trace: '
+  changetype<iovec>(iovPtr).buf_len = 7;
+  fd_write(1, iovPtr, 1, lenPtr);
+  changetype<iovec>(iovPtr).buf_len = String.UTF8.encodeUnsafe(changetype<usize>(message), message.length, bufPtr);
+  fd_write(1, iovPtr, 1, lenPtr);
+  if (n) {
+    store<u8>(bufPtr++, 0x20); // space
+    changetype<iovec>(iovPtr).buf_len = 1 + String.UTF8.encodeUnsafe(bufPtr, dtoa_stream(bufPtr, 0, a0), bufPtr);
+    fd_write(1, iovPtr, 1, lenPtr);
+    if (n > 1) {
+      changetype<iovec>(iovPtr).buf_len = 1 + String.UTF8.encodeUnsafe(bufPtr, dtoa_stream(bufPtr, 0, a1), bufPtr);
+      fd_write(1, iovPtr, 1, lenPtr);
+      if (n > 2) {
+        changetype<iovec>(iovPtr).buf_len = 1 + String.UTF8.encodeUnsafe(bufPtr, dtoa_stream(bufPtr, 0, a2), bufPtr);
+        fd_write(1, iovPtr, 1, lenPtr);
+        if (n > 3) {
+          changetype<iovec>(iovPtr).buf_len = 1 + String.UTF8.encodeUnsafe(bufPtr, dtoa_stream(bufPtr, 0, a3), bufPtr);
+          fd_write(1, iovPtr, 1, lenPtr);
+          if (n > 4) {
+            changetype<iovec>(iovPtr).buf_len = 1 + String.UTF8.encodeUnsafe(bufPtr, dtoa_stream(bufPtr, 0, a4), bufPtr);
+            fd_write(1, iovPtr, 1, lenPtr);
+          }
+        }
+      }
+    }
+    --bufPtr;
+  }
+  store<u8>(bufPtr, 0x0A); // \n
+  changetype<iovec>(iovPtr).buf_len = 1;
+  fd_write(1, iovPtr, 1, lenPtr);
+  __free(iovPtr);
+}
+
+function seed(): f64 {
+  var temp = load<u64>(0);
+  var rand: u64;
+  do {
+    random_get(0, 8); // to be sure
+    rand = load<u64>(0);
+  } while (!rand);
+  store<u64>(0, temp);
+  return reinterpret<f64>(rand);
+}
+
+export * from "bindings/wasi";
```

Comment:
With the `import "wasi"` syntax. Does this automatically provide all of the bindings, without having to import them individually?

And I assume current projects can just keep importing them individually right? :smile: This is such a great solution :smile:

Reply:
Yes, one can either enable all this stuff including stdlib-provided implementations abort/trace/seed by `import "wasi"` or use the bindings with custom implementations and command line options (similar to before), with the former implying (and re-exporting) the latter.

Reply:
In case this didn't exactly answer the question: `import * as wasi from "wasi"` is what one would do to also have all the bindings at their fingertips :)

Reply:
Oh that's perfect! Thank you! :smile: :+1: